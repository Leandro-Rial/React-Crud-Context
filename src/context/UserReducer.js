import { types } from "../types/types"

export const UserReducer = (state = {}, action) => {
    switch (action.type) {
        case types.removeUser:
            
            return {
                ...state,
                users: state.users.filter(user => {
                    return user.id !== action.payload
                })
            }
        
        case types.addUser:
            
            return {
                ...state,
                users: [action.payload, ...state.users]
            }
        
        
        case types.editUser:
            const updateUser = action.payload;
            
            const updateUsers = state.users.map(user => {
                if(user.id === updateUser.id) {
                    return updateUser
                }

                return user;
            })

            return {
                ...state,
                users: updateUsers
            }
        
    
        default:
            return state
    }
}