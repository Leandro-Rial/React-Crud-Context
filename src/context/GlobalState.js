import React, { createContext, useEffect, useReducer } from 'react';
import { types } from '../types/types';
import { UserReducer } from './UserReducer';

const initialState = {
    users: []
}

const init = () => {
    return (
        JSON.parse(localStorage.getItem("users")) || initialState
    )
}

export const UserContext = createContext(initialState);

export const UserProvider = ({ children }) => {

    const [state, dispatch] = useReducer(UserReducer, initialState, init)

    useEffect(() => {
        localStorage.setItem('users', JSON.stringify(state))
    }, [state])

    const removeUser = (id) => {
        dispatch({
            type: types.removeUser,
            payload: id
        })
    }
    
    const addUser = (user) => {
        dispatch({
            type: types.addUser,
            payload: user,
        })
    }

    const editUser = (user) => {
        dispatch({
            type: types.editUser,
            payload: user
        })
    }

    return (
        <UserContext.Provider value={{
            users: state.users,
            removeUser,
            addUser,
            editUser
        }}>
            {children}
        </UserContext.Provider>
    )
    
}