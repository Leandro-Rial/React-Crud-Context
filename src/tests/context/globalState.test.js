import { mount } from "enzyme";
import { UserContext, UserProvider } from "../../context/GlobalState";

describe('Pruebas en GlobalState', () => {
    test('debe de mostrarse correctamente', () => {
        const wrapper = mount(
            <UserContext.Provider>
                <UserProvider />
            </UserContext.Provider>
        )

        expect(wrapper).toMatchSnapshot()
    });
})
