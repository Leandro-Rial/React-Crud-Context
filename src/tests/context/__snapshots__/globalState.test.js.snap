// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GlobalState debe de mostrarse correctamente 1`] = `""`;
